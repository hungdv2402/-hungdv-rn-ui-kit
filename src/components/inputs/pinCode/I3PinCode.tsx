import React, { useMemo, useRef, ReactNode } from 'react';
import { Keyboard, Pressable, StyleSheet, Text, TextInput, View } from 'react-native';

type I3PinCodeProps = {
  isError?: boolean;
  quantity?: number;
  value: string;
  setValue(value: string): void;
  onComplete?(): void;
  setError?(isError: boolean): void;
};

const I3PinCode = ({
  isError,
  value,
  quantity = 4,
  setValue,
  onComplete,
  setError,
}: I3PinCodeProps) => {
  const inputRef = useRef<TextInput>(null);

  const renderSquares = useMemo(() => {
    const squares: ReactNode[] = [];
    const valueArray = value ? value.split('') : [];
    for (let index = 0; index < quantity; index++) {
      squares.push(
        <View
          key={index.toString()}
          style={[styles.square, isError && { borderWidth: 2, borderColor: 'red' }]}
        >
          <Text style={styles.squareText}>{valueArray[index] || ''}</Text>
        </View>
      );
    }
    return squares;
  }, [value]);

  const onChangeText = (value: string) => {
    if (value.length <= quantity) {
      setError && setError(false);
      setValue(value);
      if (value.length === quantity) {
        Keyboard.dismiss();
        onComplete && onComplete();
      }
    }
  };

  const onFocus = () => {
    inputRef.current?.focus();
    if (isError) {
      inputRef.current?.setSelection(0, quantity);
    }
  };

  return (
    <Pressable onPress={onFocus} style={styles.container}>
      {renderSquares}

      <TextInput
        value={value}
        keyboardType="numeric"
        onChangeText={onChangeText}
        caretHidden
        ref={inputRef}
        style={{ display: 'none' }}
      />
    </Pressable>
  );
};

export default I3PinCode;

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    gap: 10,
    justifyContent: 'center',
  },
  square: {
    height: 60,
    width: 50,
    borderWidth: 1,
    borderColor: '#D3D3D4',
    borderRadius: 5,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#fff',
  },
  squareText: {
    color: '#212127',
    fontSize: 20,
    fontWeight: '600',
  },
});
