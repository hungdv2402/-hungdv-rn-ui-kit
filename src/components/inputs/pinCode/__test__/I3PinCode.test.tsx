import renderer from 'react-test-renderer';
import I3PinCode from '../I3PinCode';
import React from 'react';

describe('renders I3PinCode correctly', () => {
  it('should render I3PinCode with basic properties', () => {
    const tree = renderer
      .create(
        <I3PinCode
          value={'8888'}
          setValue={jest.fn()}
          onComplete={jest.fn()}
        />
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render I3PinCode with custom errors', () => {
    const tree = renderer
      .create(
        <I3PinCode
          value={''}
          setValue={jest.fn()}
          isError={true}
          setError={jest.fn()}
        />
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render I3PinCode with custom quantity', () => {
    const tree = renderer
      .create(
        <I3PinCode
          value={'888888'}
          setValue={jest.fn()}
          quantity={6}
        />
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});