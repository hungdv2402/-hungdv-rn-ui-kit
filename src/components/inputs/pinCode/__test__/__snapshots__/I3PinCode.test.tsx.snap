// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders I3PinCode correctly should render I3PinCode with basic properties 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "flexDirection": "row",
      "gap": 10,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderColor": "#D3D3D4",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 50,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        {
          "color": "#212127",
          "fontSize": 20,
          "fontWeight": "600",
        }
      }
    >
      8
    </Text>
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderColor": "#D3D3D4",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 50,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        {
          "color": "#212127",
          "fontSize": 20,
          "fontWeight": "600",
        }
      }
    >
      8
    </Text>
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderColor": "#D3D3D4",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 50,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        {
          "color": "#212127",
          "fontSize": 20,
          "fontWeight": "600",
        }
      }
    >
      8
    </Text>
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderColor": "#D3D3D4",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 50,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        {
          "color": "#212127",
          "fontSize": 20,
          "fontWeight": "600",
        }
      }
    >
      8
    </Text>
  </View>
  <TextInput
    caretHidden={true}
    keyboardType="numeric"
    onChangeText={[Function]}
    style={
      {
        "display": "none",
      }
    }
    value="8888"
  />
</View>
`;

exports[`renders I3PinCode correctly should render I3PinCode with custom errors 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "flexDirection": "row",
      "gap": 10,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderColor": "#D3D3D4",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 50,
        },
        {
          "borderColor": "red",
          "borderWidth": 2,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "#212127",
          "fontSize": 20,
          "fontWeight": "600",
        }
      }
    />
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderColor": "#D3D3D4",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 50,
        },
        {
          "borderColor": "red",
          "borderWidth": 2,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "#212127",
          "fontSize": 20,
          "fontWeight": "600",
        }
      }
    />
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderColor": "#D3D3D4",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 50,
        },
        {
          "borderColor": "red",
          "borderWidth": 2,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "#212127",
          "fontSize": 20,
          "fontWeight": "600",
        }
      }
    />
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderColor": "#D3D3D4",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 50,
        },
        {
          "borderColor": "red",
          "borderWidth": 2,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "#212127",
          "fontSize": 20,
          "fontWeight": "600",
        }
      }
    />
  </View>
  <TextInput
    caretHidden={true}
    keyboardType="numeric"
    onChangeText={[Function]}
    style={
      {
        "display": "none",
      }
    }
    value=""
  />
</View>
`;

exports[`renders I3PinCode correctly should render I3PinCode with custom quantity 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "flexDirection": "row",
      "gap": 10,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderColor": "#D3D3D4",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 50,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        {
          "color": "#212127",
          "fontSize": 20,
          "fontWeight": "600",
        }
      }
    >
      8
    </Text>
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderColor": "#D3D3D4",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 50,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        {
          "color": "#212127",
          "fontSize": 20,
          "fontWeight": "600",
        }
      }
    >
      8
    </Text>
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderColor": "#D3D3D4",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 50,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        {
          "color": "#212127",
          "fontSize": 20,
          "fontWeight": "600",
        }
      }
    >
      8
    </Text>
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderColor": "#D3D3D4",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 50,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        {
          "color": "#212127",
          "fontSize": 20,
          "fontWeight": "600",
        }
      }
    >
      8
    </Text>
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderColor": "#D3D3D4",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 50,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        {
          "color": "#212127",
          "fontSize": 20,
          "fontWeight": "600",
        }
      }
    >
      8
    </Text>
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderColor": "#D3D3D4",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 50,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        {
          "color": "#212127",
          "fontSize": 20,
          "fontWeight": "600",
        }
      }
    >
      8
    </Text>
  </View>
  <TextInput
    caretHidden={true}
    keyboardType="numeric"
    onChangeText={[Function]}
    style={
      {
        "display": "none",
      }
    }
    value="888888"
  />
</View>
`;
